Spigot-Server-Patches/0594-Damage-with-any-cause.patch

```diff
@@ -0,0 +1,454 @@
++            case LIGHTNING:
++                return DamageSource.LIGHTNING;
++            case STARVATION:
++                return DamageSource.STARVE;
++            case POISON:
++                return CraftEventFactory.POISON;
++            case MAGIC:
++                return DamageSource.MAGIC;
++            case WITHER:
++                return DamageSource.WITHER;
++            case FALLING_BLOCK:
++                return DamageSource.FALLING_BLOCK;
++            case THORNS:
++                if (attackingEntity == null) {
++                    return DamageSource.GENERIC;
++                }
++                return DamageSource.thorns(attackingEntity);
++            case DRAGON_BREATH:
++                return DamageSource.DRAGON_BREATH;
++            case CUSTOM:
++                return DamageSource.GENERIC;
++            case FLY_INTO_WALL:
++                return DamageSource.FLY_INTO_WALL;
++            case HOT_FLOOR:
++                return DamageSource.HOT_FLOOR;
++            case CRAMMING:
++                return DamageSource.CRAMMING;
++            case DRYOUT:
++                return DamageSource.DRYOUT;
++        }
++        return DamageSource.GENERIC;
++    }
++
++}
+diff --git a/src/main/java/net/minecraft/server/DamageSource.java b/src/main/java/net/minecraft/server/DamageSource.java
+index 6fe5678cffc2487fe00c953d772f764bb37a4b11..6226935318368907a1f6d37e3f8c04e3be724f7e 100644
+--- a/src/main/java/net/minecraft/server/DamageSource.java
++++ b/src/main/java/net/minecraft/server/DamageSource.java
+@@ -1,5 +1,9 @@
+ package net.minecraft.server;
+ 
++import org.bukkit.craftbukkit.event.CraftEventFactory;
++import org.bukkit.entity.LightningStrike;
++import org.bukkit.event.entity.EntityDamageEvent;
++
+ import javax.annotation.Nullable;
+ 
+ public class DamageSource {
+@@ -150,6 +154,14 @@ public class DamageSource {
+         this.translationIndex = s;
+     }
+ 
++    public static DamageSource thorns(Entity attackingEntity) { return DamageSource.a(attackingEntity); } // Paper OBFHELPER
++
++    public static DamageSource witherSkull(EntityWitherSkull attackingEntity, IProjectile projectile) { return DamageSource.a(attackingEntity, projectile.getShooter()); } // Paper OBFHELPER
++
++    public static DamageSource fireworkRocket(EntityFireworks attackingEntity, IProjectile projectile) { return DamageSource.a(attackingEntity, projectile.getShooter()); } // Paper OBFHELPER
++
++    public static DamageSource trident(Entity attackingEntity, IProjectile projectile) { return DamageSource.a(attackingEntity, projectile.getShooter()); } // Paper OBFHELPER
++
+     @Nullable
+     public Entity j() {
+         return this.getEntity();
+@@ -226,4 +238,52 @@ public class DamageSource {
+     public Vec3D w() {
+         return null;
+     }
++
++    // Paper - start
++    public static EntityDamageEvent.DamageCause toDamageCause(DamageSource source, @Nullable Entity damager, Entity damagee) {
```

Comment:
if we're adding all of this, any reason this isn't in the other class?

Reply:
Is this a request to move it elsewhere? 
Its just there because i did not find better place.